include/kernel.h

```diff
@@ -0,0 +1,531 @@
+	mb; \
+})
+
+/**
+ * @brief Free a heap mailbox when it's not needed anymore.
+ *
+ * @note Do not call this function on a mailbox allocated with
+ *       #kmbox_new_stack. As the mailbox is on the stack, there is
+ *       no need to explicitly free it.
+ */
+void kmbox_free(kmbox_t *mbox);
+
+/** @brief Return true if the mailbox is empty */
+bool kmbox_empty(kmbox_t *mbox);
+
+/** @brief Return true if the mailbox is full */
+bool kmbox_full(kmbox_t *mbox);
+
+/** 
+ * @brief Try sending a mail to a mailbox. 
+ *
+ * This function tries to send a mail to a mailbox. If the
+ * mailbox is full, it returns false without blocking.
+ *
+ * @param[in] mbox The mailbox to send the mail to
+ * @param[in] mail The mail to send (must not be NULL)
+ *
+ * @return true if the mail was sent, false if the mailbox was full
+ */
+bool kmbox_try_send(kmbox_t *mbox, void *mail);
+
+/** 
+ * @brief Send a mail to a mailbox. 
+ *
+ * This function sends a mail to a mailbox. If the
+ * mailbox is full, it will block until another thread
+ * receive a pending mail, so that the mailbox has space
+ * for the new mail.
+ *
+ * @param[in] mbox The mailbox to send the mail to
+ * @param[in] mail The mail to send (must not be NULL)
+ *
+ */
+void kmbox_send(kmbox_t *mbox, void *mail);
+
+/** 
+ * @brief Try receiving a mail from a mailbox. 
+ *
+ * This function tries to receive a mail to a mailbox. If the
+ * mailbox is empty, it will return NULL without blocking.
+ *
+ * @param[in] mbox The mailbox to receive the mail from
+ *
+ * @return the received mail, or NULL if the mailbox was empty
+ *
+ */
+void* kmbox_try_recv(kmbox_t *mbox);
+
+/** 
+ * @brief Receive a mail from a mailbox. 
+ *
+ * This function receives a mail from a mailbox. If the
+ * mailbox is empty, it will block until a mail is available.
+ *
+ * @param[in] mbox The mailbox to receive the mail from
+ *
+ * @return the received mail
+ *
+ */
+void* kmbox_recv(kmbox_t *mbox);
+
+
+/**
+ * @brief Attach an event to a mailbox, so that a mail arrives when the event
+ *        is triggered.
+ *
+ * This functions connects a mailbox to the specified event, so that
+ * when the event is triggered, a mail is posted to the
+ * mailbox. The thread can then poll the mailbox to see when the
+ * event is arrived. The received mail will be the event itself.
+ *
+ * This function is useful for threads that need to react to multiple
+ * events and/or handle messages from other threads. If you just
+ * want to wait for a single interrupt, it is easier to simply
+ * call #thread_wait_event.
+ *
+ * Example:
+ *
+ *     kmbox_t *mbox = kmbox_new_stack(4);
+ *     kmbox_attach_event(mbox, &KEVENT_IRQ_SP);
+ *     kmbox_attach_event(mbox, &KEVENT_IRQ_DP);
+ *
+ *     void *mail;
+ *     while ((mail = kmbox_recv(mbox)))
+ *     {
+ *        if (mail == &KEVENT_IRQ_SP)
+ *        {
+ *            [...]
+ *        }
+ *        else if (mail == &KEVENT_IRQ_DP)
+ *        {
+ *            [...]
+ *        }
+ *     }
```

Comment:
Not to speak for every programmer, but I like this a lot. Or, it feels comfortably straightforward if someone's making a game by themselves, but also easy to put into something larger.